### PR TITLE
Fix conflict between param and exact path

### LINKED ISSUE #2682: [Bug] Mixed params and exact paths can cause a 404
> ```
> Router1 "/get/test/abc/"
> Router2 "/get/:param/abc/"
> ```
> I tests these two routers with some requests:
> /get/test/abc/ -> Router1
> /get/xyz/abc/ -> Router2
> /get/tt/abc/ -> 404(Expect Router2)
> So, if there are some param starts with the prefix 't' in this example, it will not get the right router.

_Originally posted by @Tevic in https://github.com/gin-gonic/gin/issues/2663#issuecomment-814575677_

### LINKED ISSUE #2696: Matching prefixes among multiple routers causes 404
I'm not particularly experienced with Gin, so I might well have gotten something wrong here. However, the following behavior appears strange to me. Consider this small example:

```go
package main

import (
	"net/http"
	"github.com/gin-gonic/gin"
)

func main() {
	r := gin.New()

	// useless because of https://github.com/golang/go/blob/a7e16abb22f1b249d2691b32a5d20206282898f2/src/net/http/fs.go#L587, but anyway ...
	r.StaticFile("/index.html", "./public/index.html")
	r.StaticFile("/robots.txt", "./public/robots.txt")
	r.GET("/:id", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{"status": "success"})
	})
	
	r.Run()
}
```

## Expected vs. actual behavior
```bash
$ curl http://localhost:8080/aaaa # -> shall return 200, does return 200
$ curl http://localhost:8080/iaaa # -> shall return 200, does return 404
$ curl http://localhost:8080/raaa # -> shall return 200, does return 404
```

If I remove the `robots.txt` handler, the third request will be `200` instead. Looks like if my request's path has a prefix, which is also a prefix of one of the other handlers, things get messed up.

Might be a duplicate of https://github.com/gin-gonic/gin/issues/2682.

### PR DESCRIPTION
**Pull request by g1eny0ung**
Signed-off-by: Yue Yang <g1enyy0ung@gmail.com>

Close #2682.
Close #2696.

This PR solves the above issues.

~~For a detailed description, see https://github.com/gin-gonic/gin/issues/2682#issuecomment-824579980.~~

### CODE REVIEW

**Comment by g1eny0ung** [comment]
@thinkerou Would you please take a look if you are free? Very thx. 😃

**Comment by appleboy** [comment]
@g1eny0ung Can you also send the PR to https://github.com/julienschmidt/httprouter?

**Comment by g1eny0ung** [comment]
> @g1eny0ung Can you also send the PR to https://github.com/julienschmidt/httprouter?

Because the PR https://github.com/julienschmidt/httprouter/pull/329 still not be merged, so it's needed to wait for it or let @rw-access update the related code.

File: tree.go [hunk 1]
```diff
@@ -411,8 +411,14 @@ walk: // Outer loop for walking the tree
 				idxc := path[0]
 				for i, c := range []byte(n.indices) {
 					if c == idxc {
-						n = n.children[i]
-						continue walk
+						childWillWalkTo := n.children[i]
+
+						if strings.Split(childWillWalkTo.path, "/")[0] != strings.Split(path, "/")[0] && strings.HasPrefix(n.children[len(n.children)-1].path, ":") {
```
Comment (rw-access) [inline tree.go:416]: I think this will cause allocations which are currently avoided and could cause a slowdown

File: tree_test.go [hunk 2]
```diff
@@ -169,6 +170,7 @@ func TestTreeWildcard(t *testing.T) {
 		{"/search/", false, "/search/", nil},
 		{"/search/someth!ng+in+ünìcodé", false, "/search/:query", Params{Param{Key: "query", Value: "someth!ng+in+ünìcodé"}}},
 		{"/search/someth!ng+in+ünìcodé/", true, "", Params{Param{Key: "query", Value: "someth!ng+in+ünìcodé"}}},
+		{"/search/gin", false, "/search/:query", Params{Param{"query", "gin"}}},
```
Comment (rw-access) [inline tree_test.go:173]: I think this is more complicated than the proposed fix above.

Try adding some more routes to cover some bases like these.  Because of the tree structure, this doesn't look like it will incorrectly handle when these routes exist

-  `/search/gin-gonic`
- `/search/google`
- `/search/:param`

If you send in `/search/gaga` which route is hit? The last one?

I think we need a lot of tests, which I failed to do on the original PR.

I'm sorry that I haven't had much time to tackle this, but I believe that the best answer involves some lightweight backtracking that can be done cheaply.

**Comment by appleboy** [comment]
@g1eny0ung Please provide the benchmark result like https://github.com/gin-gonic/gin/pull/2663#issuecomment-811665303 and https://github.com/gin-gonic/gin/pull/2663#issuecomment-813111923

**Comment by g1eny0ung** [comment]
@rw-access @appleboy Thanks for the notes mentioned. I will try to add more tests and the benchmark later to make sure the change code is stable.

File: tree.go (hunk 1 above)
Comment (g1eny0ung) [reply tree.go:416]: Agree with u. If we add some conditions to check the param in advance, there should be a little slowdown than before.

**Comment by g1eny0ung** [comment]
@appleboy After @rw-access noted (backtracking) and some thoughts 🤔, I update the code to save a temporary `skipped` value when the children have a param node at the last.

The skipped contains the current node path and a param node (same as the current node except no indices). When a param path has the same prefix with the exact path, the prefix tree will work as before, keep searching downwards, but it will fail. 

Then the follow-up logic will check the `skipped` value is not `nil` to perform a backtracking search. The param node in `skipped` will be set to the current node. Because it has no **indices**, it will jump the non-wildcard children check, go directly to the place of param judgment.

Compared to before, I have added more tests, if there are omissions, please help me to add. 😆

The benchmarks:

**master** - https://travis-ci.com/github/justttype/go-http-routing-benchmark/builds/224357913
**this PR** - https://travis-ci.com/github/justttype/go-http-routing-benchmark/builds/224455608

The `benchstat` report:

```
name              old time/op    new time/op    delta
Gin_Param           66.6ns ± 0%    67.6ns ± 0%   ~     (p=1.000 n=1+1)
Gin_Param5           117ns ± 0%     112ns ± 0%   ~     (p=1.000 n=1+1)
Gin_Param20          303ns ± 0%     298ns ± 0%   ~     (p=1.000 n=1+1)
Gin_ParamWrite       123ns ± 0%     119ns ± 0%   ~     (p=1.000 n=1+1)
Gin_GithubStatic    77.4ns ± 0%    90.0ns ± 0%   ~     (p=1.000 n=1+1)
Gin_GithubParam      133ns ± 0%     156ns ± 0%   ~     (p=1.000 n=1+1)
Gin_GithubAll       26.8µs ± 0%    31.1µs ± 0%   ~     (p=1.000 n=1+1)
Gin_GPlusStatic     61.7ns ± 0%    66.8ns ± 0%   ~     (p=1.000 n=1+1)
Gin_GPlusParam      82.5ns ± 0%    92.2ns ± 0%   ~     (p=1.000 n=1+1)
Gin_GPlus2Params     103ns ± 0%     116ns ± 0%   ~     (p=1.000 n=1+1)
Gin_GPlusAll        1.13µs ± 0%    1.29µs ± 0%   ~     (p=1.000 n=1+1)
Gin_ParseStatic     62.3ns ± 0%    65.5ns ± 0%   ~     (p=1.000 n=1+1)
Gin_ParseParam      71.4ns ± 0%    74.3ns ± 0%   ~     (p=1.000 n=1+1)
Gin_Parse2Params    84.4ns ± 0%    84.7ns ± 0%   ~     (p=1.000 n=1+1)
Gin_ParseAll        2.01µs ± 0%    2.16µs ± 0%   ~     (p=1.000 n=1+1)
Gin_StaticAll       18.5µs ± 0%    23.7µs ± 0%   ~     (p=1.000 n=1+1)

name              old alloc/op   new alloc/op   delta
Gin_Param            0.00B          0.00B        ~     (all equal)
Gin_Param5           0.00B          0.00B        ~     (all equal)
Gin_Param20          0.00B          0.00B        ~     (all equal)
Gin_ParamWrite       0.00B          0.00B        ~     (all equal)
Gin_GithubStatic     0.00B          0.00B        ~     (all equal)
Gin_GithubParam      0.00B          0.00B        ~     (all equal)
Gin_GithubAll        0.00B          0.00B        ~     (all equal)
Gin_GPlusStatic      0.00B          0.00B        ~     (all equal)
Gin_GPlusParam       0.00B          0.00B        ~     (all equal)
Gin_GPlus2Params     0.00B          0.00B        ~     (all equal)
Gin_GPlusAll         0.00B          0.00B        ~     (all equal)
Gin_ParseStatic      0.00B          0.00B        ~     (all equal)
Gin_ParseParam       0.00B          0.00B        ~     (all equal)
Gin_Parse2Params     0.00B          0.00B        ~     (all equal)
Gin_ParseAll         0.00B          0.00B        ~     (all equal)
Gin_StaticAll        0.00B          0.00B        ~     (all equal)

name              old allocs/op  new allocs/op  delta
Gin_Param             0.00           0.00        ~     (all equal)
Gin_Param5            0.00           0.00        ~     (all equal)
Gin_Param20           0.00           0.00        ~     (all equal)
Gin_ParamWrite        0.00           0.00        ~     (all equal)
Gin_GithubStatic      0.00           0.00        ~     (all equal)
Gin_GithubParam       0.00           0.00        ~     (all equal)
Gin_GithubAll         0.00           0.00        ~     (all equal)
Gin_GPlusStatic       0.00           0.00        ~     (all equal)
Gin_GPlusParam        0.00           0.00        ~     (all equal)
Gin_GPlus2Params      0.00           0.00        ~     (all equal)
Gin_GPlusAll          0.00           0.00        ~     (all equal)
Gin_ParseStatic       0.00           0.00        ~     (all equal)
Gin_ParseParam        0.00           0.00        ~     (all equal)
Gin_Parse2Params      0.00           0.00        ~     (all equal)
Gin_ParseAll          0.00           0.00        ~     (all equal)
Gin_StaticAll         0.00           0.00        ~     (all equal)
```

Thank you all for your reviews.

File: tree_test.go [hunk 3]
```diff
@@ -135,13 +135,16 @@ func TestTreeWildcard(t *testing.T) {
 
 	routes := [...]string{
 		"/",
-		"/cmd/:tool/:sub",
 		"/cmd/:tool/",
+		"/cmd/:tool/:sub",
```
Comment (rw-access) [inline tree_test.go:139]: Thanks for all the updates, I really appreciate it.

I think we could use another test to check edge case to check:


if you have routes like this set up:
* `/cmd/whoami/foo`
* `/cmd/:tool/`
* `/cmd/:tool/:subtool`

What happens if someone hits the route:
* `/cmd/whoami/bar`?

It should cause a 404 and not hit `/cmd/:tool/:subtool`.

Basically, whenever the tree is exhausted or a the end of a path segment is found, the backtrack should be invalidated. Once you hit `/cmd/whoami/`, it shouldn't be possible to fall back to `/cmd/tool/`.

File: tree.go [hunk 4]
```diff
@@ -411,6 +418,21 @@ walk: // Outer loop for walking the tree
 				idxc := path[0]
 				for i, c := range []byte(n.indices) {
 					if c == idxc {
+						if strings.HasPrefix(n.children[len(n.children)-1].path, ":") {
```
Comment (rw-access) [inline tree.go:421]: I think you want to check if the last child is of type Param, right?

Comment (rw-access) [reply tree.go:421]: and then when you need to backtrack, then you can go to the Param child directly?

File: tree_test.go (hunk 3 above)
Comment (g1eny0ung) [reply tree_test.go:139]: This is the problem.

For example, `/cmd/who/` will not hit `/cmd/:tool/` because it has the same prefix as `/cmd/whoami/foo`, but it should match.

File: tree.go (hunk 4 above)
Comment (g1eny0ung) [reply tree.go:421]: So I think is needed to save the Param child to a place? Otherwise, I will not be able to find it because the walk loop will replace the previous node with the current.

File: tree_test.go (hunk 3 above)
Comment (g1eny0ung) [reply tree_test.go:139]: Know what you mean. I will think about this situation.

Comment (g1eny0ung) [reply tree_test.go:139]: But what still let me confused is: with the above routes, whether `/cmd/whoami/bar` hit or not, seems they are meaningful?

According to my understanding and my actual use, if I don’t have an exact route definition like `/cmd/whoami/foo`, then `/cmd/whoami/bar` needs to be matched to the wildcard route.

**Review by daheige — APPROVED** [review]

**Comment by appleboy** [comment]
@rw-access needs your feedback.

**Comment by vaimr** [comment]
It's very important fix for everybody!

**Comment by daheige** [comment]
nice!


------------------ Original ------------------
From: Denis Saponenko ***@***.***&gt;
Date: Wed,May 12,2021 7:13 PM
To: gin-gonic/gin ***@***.***&gt;
Cc: heige ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [gin-gonic/gin] Fix conflict between param and exact path (#2706)





 
It's very important fix for everybody!

**Comment by daheige** [comment]
nice,great job!


------------------ Original ------------------
From: Denis Saponenko ***@***.***&gt;
Date: Wed,May 12,2021 7:13 PM
To: gin-gonic/gin ***@***.***&gt;
Cc: heige ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [gin-gonic/gin] Fix conflict between param and exact path (#2706)





 
It's very important fix for everybody!

File: tree.go [hunk 5]
```diff
@@ -411,6 +418,21 @@ walk: // Outer loop for walking the tree
...
+							skipped = &skip{
+								path: prefix + path,
+								paramNode: &node{
```
Comment (rw-access) [inline tree.go:424]: why not just make this data structure take `n`?

File: tree.go [hunk 6]
```diff
@@ -411,6 +418,21 @@ walk: // Outer loop for walking the tree
...
+									path:      n.path,
+									wildChild: n.wildChild,
+									nType:     n.nType,
+									priority:  n.priority,
+									children:  n.children,
+									handlers:  n.handlers,
+									fullPath:  n.fullPath,
+								},
```
Comment (rw-access) [inline tree.go:432]: would this work?
```suggestion
								paramNode: n
```

**Review by rw-access — APPROVED** [review]
LGTM. Thanks for the fixes

File: tree.go (hunk 6 above)
Comment (g1eny0ung) [reply tree.go:432]: Because this operation removes indices to prevent prefix checking.

File: tree.go (hunk 5 above)
Comment (g1eny0ung) [reply tree.go:424]: Because this operation removes indices to prevent prefix checking.

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
I will bump the new version v1.7.2 if the PR has been merge. Waiting for @thinkerou approval.

**Review by thinkerou — APPROVED** [review]
lgtm

**Comment by appleboy** [comment]
Hi All,

We already released https://github.com/gin-gonic/gin/releases/tag/v1.7.2 version, just include the changes.
